File: src/rex/states/listening.py
"""
Listening state handler.

Handles speech capture and transcription, either after wake word
detection or during follow-up conversation.
"""

from typing import TYPE_CHECKING

from core.state_machine import ConversationState, StateHandler, StateResult

if TYPE_CHECKING:
    from core.context import AppContext
    from stt import Transcriber
    from wake_word import WakeWordListener

# Phrases that end the conversation
STOP_PHRASES = ("stop", "nevermind", "never mind", "cancel", "forget it")


class ListeningHandler(StateHandler):
    """
    Handler for the LISTENING state.

    Processes captured audio or waits for follow-up speech,
    then transcribes and handles special commands.
    """

    def __init__(
        self,
        listener: "WakeWordListener",
        transcriber: "Transcriber",
    ):
        self._listener = listener
        self._transcriber = transcriber
        self._audio = None
        self._is_wake_word_trigger = False

    @property
    def state(self) -> ConversationState:
        return ConversationState.LISTENING

    def enter(self, ctx: "AppContext", data: dict | None = None) -> None:
        """Store audio data if provided."""
        if data:
            self._audio = data.get("audio")
            self._is_wake_word_trigger = data.get("is_wake_word_trigger", False)
        else:
            self._audio = None
            self._is_wake_word_trigger = False

    def process(self, ctx: "AppContext") -> StateResult:
        """
        Process audio or capture follow-up speech.

        Returns:
            PROCESSING if transcription successful
            WAITING_FOR_WAKE_WORD if timeout or stop command
        """
        # If no audio provided (follow-up mode), listen for speech
        if self._audio is None:
            # Play ready tone BEFORE waiting for speech so user knows Rex is listening
            ctx.audio_manager.play_listening_tone()
            print("🎤 Listening for response...")
            self._audio = self._listener.listen_for_speech(
                timeout=ctx.settings.listening_timeout, play_tones=True
            )

            if self._audio is None:
                print("⏱️ No response received, ending conversation.")
                # Play done tone to indicate we're no longer listening
                ctx.audio_manager.play_done_tone()
                return StateResult(next_state=ConversationState.WAITING_FOR_WAKE_WORD)

        # Transcribe the audio
        strip_wake_word = self._is_wake_word_trigger
        transcription = self._transcriber.transcribe(self._audio, strip_wake_word=strip_wake_word)

        if not transcription:
            # Empty transcription - go back to appropriate state
            if ctx.is_in_conversation():
                # Stay listening for follow-up
                return StateResult(
                    next_state=ConversationState.LISTENING,
                    data={"audio": None, "is_wake_word_trigger": False},
                )
            else:
                return StateResult(next_state=ConversationState.WAITING_FOR_WAKE_WORD)

        print(f"\n💬 You said: {transcription}\n")

        # Handle special commands
        normalized = transcription.strip().lower()

        # Timer stop command (works anytime)
        if normalized in ("stop", "stop the timer"):
            if ctx.timer_manager and ctx.timer_manager.stop_any_ringing():
                print("🔕 Timer alarm stopped.")
            return StateResult(next_state=ConversationState.WAITING_FOR_WAKE_WORD)

        # Stop phrases end conversation (only during follow-up)
        if ctx.is_in_conversation() and normalized in STOP_PHRASES:
            return StateResult(next_state=ConversationState.WAITING_FOR_WAKE_WORD)

        # Pass transcription to processing state
        return StateResult(
            next_state=ConversationState.PROCESSING,
            data={"transcription": transcription},
        )

    def exit(self, ctx: "AppContext") -> None:
        """Clear audio data."""
        self._audio = None
        self._is_wake_word_trigger = False

File: internal/states/listening.go
package states

import (
	"context"
	"log/slog"
)

// ListeningHandler implements the LISTENING state.
// It captures or receives audio, transcribes it, and routes to the next state.
type ListeningHandler struct {
	deps             *Deps
	audio            []int16
	isWakeWordTrigger bool
}

// NewListeningHandler creates a ListeningHandler.
func NewListeningHandler(deps *Deps) *ListeningHandler {
	return &ListeningHandler{deps: deps}
}

func (h *ListeningHandler) State() int { return StateListening }

// Enter stores audio passed from a previous state (e.g. wake-word capture or interruption).
func (h *ListeningHandler) Enter(data map[string]any) {
	h.audio = nil
	h.isWakeWordTrigger = false
	if data == nil {
		return
	}
	if a, ok := data["audio"]; ok && a != nil {
		if samples, ok := a.([]int16); ok {
			h.audio = samples
		}
	}
	if v, ok := data["is_wake_word_trigger"]; ok {
		if b, ok := v.(bool); ok {
			h.isWakeWordTrigger = b
		}
	}
}

// Process transcribes existing audio or captures new speech, then decides
// the next state based on the transcription content.
func (h *ListeningHandler) Process(ctx context.Context) StateResult {
	// If no audio was provided (follow-up mode), listen for speech.
	if h.audio == nil {
		h.deps.Audio.PlayListeningTone()
		slog.Info("listening for response")

		timeout := h.deps.Settings.ListeningTimeout
		audio, err := h.deps.Listener.ListenForSpeech(ctx, timeout, true)
		if err != nil {
			slog.Error("listen error", "err", err)
			h.deps.Audio.PlayDoneTone()
			return StateResult{NextState: StateWaitingForWakeWord}
		}
		if audio == nil {
			slog.Info("no response received, ending conversation")
			h.deps.Audio.PlayDoneTone()
			return StateResult{NextState: StateWaitingForWakeWord}
		}
		h.audio = audio
	}

	// Transcribe the captured audio.
	transcription, err := h.deps.Transcriber.Transcribe(h.audio, h.isWakeWordTrigger)
	if err != nil {
		slog.Error("transcription error", "err", err)
		return StateResult{NextState: StateWaitingForWakeWord}
	}

	if transcription == "" {
		if h.deps.IsInConversation() {
			return StateResult{
				NextState: StateListening,
				Data:      map[string]any{"audio": nil, "is_wake_word_trigger": false},
			}
		}
		return StateResult{NextState: StateWaitingForWakeWord}
	}

	slog.Info("transcribed", "text", transcription)

	// Handle "stop" / "stop the timer" commands.
	normalized := trimLower(transcription)
	if normalized == "stop" || normalized == "stop the timer" {
		if h.deps.Timers != nil && h.deps.Timers.StopAnyRinging() {
			slog.Info("timer alarm stopped")
		}
		return StateResult{NextState: StateWaitingForWakeWord}
	}

	// Stop phrases end the conversation during follow-up.
	if h.deps.IsInConversation() && IsStopPhrase(normalized) {
		return StateResult{NextState: StateWaitingForWakeWord}
	}

	return StateResult{
		NextState: StateProcessing,
		Data:      map[string]any{"transcription": transcription},
	}
}

// Exit clears transient audio data.
func (h *ListeningHandler) Exit() {
	h.audio = nil
	h.isWakeWordTrigger = false
}

